api/organisations/tasks.py

```diff
@@ -73,3 +86,85 @@ def finish_subscription_cancellation():
     ):
         subscription.organisation.cancel_users()
         subscription.save_as_free_subscription()
+
+
+def send_admin_api_usage_notification(
+    organisation: Organisation, matched_threshold: int
+) -> None:
+    """
+    Send notification to admins that the API has breached a threshold.
+    """
+    recipient_list = list(
+        FFAdminUser.objects.filter(
+            userorganisation__organisation=organisation,
+            userorganisation__role=OrganisationRole.ADMIN,
+        ).values_list("email", flat=True)
+    )
+
+    context = {
+        "organisation": organisation,
+        "matched_threshold": matched_threshold,
+    }
+    send_mail(
+        subject=f"Flagsmith API use has reached {matched_threshold}%",
+        message=render_to_string("organisations/api_usage_notification.txt", context),
+        from_email=settings.DEFAULT_FROM_EMAIL,
+        recipient_list=recipient_list,
+        fail_silently=True,
+    )
+
+    OranisationAPIUsageNotification.objects.create(
+        organisation=organisation,
+        percent_usage=matched_threshold,
+        notified_at=timezone.now(),
+    )
+
+
+def _handle_api_usage_notifications(organisation: Organisation):
+    subscription_cache = organisation.subscription_information_cache
+    billing_starts_at = subscription_cache.current_billing_term_starts_at
+    now = timezone.now()
+
+    # Truncate to the closest active month to get start of current period..
+    month_delta = relativedelta(now, billing_starts_at).months
+    period_starts_at = relativedelta(months=month_delta) + billing_starts_at
+
+    days = relativedelta(now, period_starts_at).days
+    api_usage = get_current_api_usage(organisation.id, f"{days}d")
+
+    api_usage_percent = int(100 * api_usage / subscription_cache.allowed_30d_api_calls)
+
+    matched_threshold = None
+    for threshold in API_USAGE_ALERT_THRESHOLDS:
+        if threshold > api_usage_percent:
+            break
+
+        matched_threshold = threshold
+
+    if OranisationAPIUsageNotification.objects.filter(
+        notified_at__gt=period_starts_at,
+        percent_usage=matched_threshold,
+    ).exists():
+        # Already sent the max notification level so don't resend.
+        return
+
+    send_admin_api_usage_notification(organisation, matched_threshold)
+
+
+@register_recurring_task(
+    run_every=timedelta(hours=12),
+)
+def handle_api_usage_notifications():
+    for organisation in Organisation.objects.filter(
+        subscription_information_cache__current_billing_term_starts_at__isnull=False,
+        subscription_information_cache__current_billing_term_ends_at__isnull=False,
+    ).select_related(
+        "subscription_information_cache",
+    ):
```

Comment:
I'm a little bit unsure how / if we plan on backfilling data for existing organisations that don't have these billing_term attributes.

Reply:
Well there are two different cases to consider:

1. Existing ChargeBee subscribers.
2. Non-ChargeBee subscribers.

For the first case we rely on subscriptions coming in regularly enough that the billing terms should be added in eventually and should be kept in sync often enough as well.

For the second case, which I think is mostly non-SaaS users that have set up outside arrangements the current approach will need some sort of addition in order for it to work. But even there, I'm not sure what the source of truth is for those arrangements. Where do we record their billing dates and could we work that in?

Reply:
As discussed, this data will be updated on any webhook from chargebee that includes the subscription. Worst case scenario is that an annual subscription doesn't get updated for 1 year, but we can manually trigger a webhook from chargebee to update this if necessary. Under normal circumstances, this should only ever be 1 month before we have all paying subscribers up to date.